### PR TITLE
Add RPM's build and deploy (in travis)

### PR DESCRIPTION
**Pull request by SudoBobo**

### CODE REVIEW

File: luacheck/luacheck.lua [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env tarantool
+package.path = package.path .. ";/usr/share/tarantool/luacheck/?/init.lua;"
+package.path = package.path .. ";/usr/share/tarantool/luacheck/?.lua;"
```
Comment (Totktonada) [inline luacheck/luacheck.lua]: It will not with rocks.

File: rpm/luacheck.spec [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+%prep
+%setup -q -n luacheck-%{version}
+
+%install
+cd ..
+rm -rf luacheck
+mkdir luacheck
+cp -r luacheck-%{version}/* ./luacheck
+rm -rf luacheck-%{version}
```
Comment (Totktonada) [inline rpm/luacheck.spec]: What do you do?

File: .travis.yml [hunk 3]
```diff
@@ -1,41 +1,69 @@
-language: c
```
Comment (Totktonada) [inline .travis.yml]: Why do you removed existing CI?

File: rpm/luacheck.spec (hunk 2 above)
Comment (SudoBobo) [reply rpm/luacheck.spec]: Removed this.

File: luacheck/luacheck.lua (hunk 1 above)
Comment (SudoBobo) [reply luacheck/luacheck.lua]: Can you explain please?

Comment (Totktonada) [reply luacheck/luacheck.lua]: will not →will not work

I mean, install path for rocks is other, it containts `.rocks`. Moreover, install path for modules can be, say, /usr/local/…

File: .travis.yml (hunk 3 above)
Comment (SudoBobo) [reply .travis.yml]: Moved it back.

File: luacheck/luacheck.lua (hunk 1 above)
Comment (SudoBobo) [reply luacheck/luacheck.lua]: Removed absolute paths.

File: install.lua [hunk 4]
```diff
@@ -1,7 +1,8 @@
-#!/usr/bin/env lua
+#!/usr/bin/env tarantool
 local dirsep = package.config:sub(1, 1)
 local is_windows = dirsep == "\\"
-package.path = "src" .. dirsep .. "?.lua"
+package.path = package.path .. ";src" .. dirsep .. "?.lua;"
+require("luacheck.init")
```
Comment (Totktonada) [inline install.lua]: What is purpose of this require?

File: install.lua [hunk 5]
```diff
@@ -80,10 +83,11 @@ if is_windows then
 "%s" -e "package.path=[[%%~dp0..\src\?.lua;%%~dp0..\src\?\init.lua;]]..package.path" "%%~dp0luacheck.lua" %%*
 ]=]):format(args.lua))
 else
-   fh:write(([=[
+    fh:write(([=[
 #!/bin/sh
+cd "%s"
```
Comment (Totktonada) [inline install.lua]: So, file paths will be treated relative to the args.source directory?

File: rpm/luacheck.spec [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+Summary: A static analyzer and a linter for Lua
+Group: Development/Tools
+License: MIT
+URL: https://github.com/mpeterv/luacheck
+Source0: luacheck-%{version}.tar.gz
+
+BuildArch: noarch
+
+BuildRequires: tarantool >= 1.7.2.0
+Requires: tarantool >= 1.7.2.0
```
Comment (Totktonada) [inline rpm/luacheck.spec]: Suggested to use 1.9, because it is minimal version we support.

File: rpm/luacheck.spec [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+and a few other typical problems within Lua programs.
+
+%prep
+%setup -q -n luacheck-%{version}
+
+%install
+mv ./bin/luacheck.lua luacheck.lua
+tarantool install.lua . %{_prefix}/share/tarantool/luacheck
+mkdir luacheck
+cp -r ./src/luacheck/* ./luacheck
```
Comment (Totktonada) [inline rpm/luacheck.spec]: > mkdir luacheck
> cp -r ./src/luacheck/* ./luacheck

Is equivalent to the following?

> mv ./src/luacheck .

File: rpm/luacheck.spec [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+mkdir -p %{buildroot}%{_prefix}/share/tarantool/luacheck
+cp -R . %{buildroot}%{_prefix}/share/tarantool/luacheck/
+
+%files
+%{_prefix}/share/tarantool/luacheck
+%{_bindir}/luacheck
+
+%changelog
+
+* Tue Jul 10 2018 Ivan Koptelov <ivan.koptelov@tarantool.org> 0.0.1-1
```
Comment (Totktonada) [inline rpm/luacheck.spec]: 0.0.0 is Version above, but 0.0.1 here? What is real version should be?

File: spec/cli_spec.lua [hunk 9]
```diff
@@ -1,5 +1,6 @@
 local fs = require "luacheck.fs"
 local utils = require "luacheck.utils"
+
```
Comment (Totktonada) [inline spec/cli_spec.lua]: Redundant change.

File: src/luacheck/main.lua [hunk 10]
```diff
@@ -2,7 +2,7 @@ local argparse = require "luacheck.argparse"
 local builtin_standards = require "luacheck.builtin_standards"
 local config = require "luacheck.config"
 local fs = require "luacheck.fs"
-local luacheck = require "luacheck"
+local luacheck = require "luacheck.init"
```
Comment (Totktonada) [inline src/luacheck/main.lua]: Are not that equivalent (here and below)?

**Review by Totktonada — COMMENTED** [review]
Please, consider comments below.

File: .travis.yml [hunk 11]
```diff
@@ -58,10 +58,8 @@ script:
         path/to/luacheck/bin/luacheck spec/*.lua
         mv src2 src
     else
-        VERSION=`git describe --tags | sed 's/\(.*\)-.*/\1/' | sed 's/[^0-9-]*//g' | \
-            cut -c 2- | sed 's#-#\.#g' | sed 's/\.[^.]*$//'`
-        RELEASE=`git describe --tags | sed 's/\(.*\)-.*/\1/' | sed 's/[^0-9-]*//g' | \
-            cut -c 2- | sed 's#-#\.#g' | cut -d '.' -f4`
+        VERSION=`git describe --tags | sed 's/\(.*\)-.*/\1/' | cut -d '-' -f-1`
+        RELEASE=`git describe --tags | sed 's/\(.*\)-.*/\1/' | cut -d '-' -f2`
```
Comment (Totktonada) [inline .travis.yml]: Please, update variables and comment above likewise we discussed in the lulpeg PR before.

File: install.lua (hunk 4 above)
Comment (SudoBobo) [reply install.lua]: It's an error. Removed.

File: install.lua (hunk 5 above)
Comment (SudoBobo) [reply install.lua]: It was a mistake. I removed the line. Now it works just fine due to package.path, where paths to luacheck sources is stored. These paths are added to package.path during package building, also it is original luacheck way, so I think it's ok.

File: rpm/luacheck.spec (hunk 7 above)
Comment (SudoBobo) [reply rpm/luacheck.spec]: Yes, it is. Thank you, now it's more simple.

File: rpm/luacheck.spec (hunk 8 above)
Comment (SudoBobo) [reply rpm/luacheck.spec]: Actually we don't use it at all. Packpack set it to VERSION during build or, if it is not possible to get VERSION (e.g 'v0.1) , version is set to 0.0.1
I prefer to set it to the current tag for information needs.

File: spec/cli_spec.lua (hunk 9 above)
Comment (SudoBobo) [reply spec/cli_spec.lua]: Removed.

File: src/luacheck/main.lua (hunk 10 above)
Comment (SudoBobo) [reply src/luacheck/main.lua]: It's not. There is also luacheck.lua. In build time we change project structure and an error occurs.

File: rpm/luacheck.spec (hunk 8 above)
Comment (Totktonada) [reply rpm/luacheck.spec]: So, 0.0.0 above is just boilerplate? Looks weird…

File: src/luacheck/main.lua (hunk 10 above)
Comment (Totktonada) [reply src/luacheck/main.lua]: Can we discuss why it didn’t occur for building againts lua?

File: .travis.yml (hunk 11 above)
Comment (SudoBobo) [reply .travis.yml]: Ok, done.

File: rpm/luacheck.spec (hunk 6 above)
Comment (SudoBobo) [reply rpm/luacheck.spec]: Sure.

File: rpm/luacheck.spec (hunk 8 above)
Comment (SudoBobo) [reply rpm/luacheck.spec]: We can't remove because it breaks down `rpmbuild`

File: src/luacheck/main.lua (hunk 10 above)
Comment (SudoBobo) [reply src/luacheck/main.lua]: Didn't understand you. I'd like to discuss verbally.

Comment (SudoBobo) [reply src/luacheck/main.lua]: Discussed verbally. Decided to use `require "luacheck" ` and therefore keep original requires. Also decided to remove /bin/luacheck.lua and call src/luacheck/main.lua from luacheck.sh instead.

Comment (Totktonada) [reply src/luacheck/main.lua]: Yep. main.lua shoudl contain shebang and has executable bit.

Comment (Totktonada) [reply src/luacheck/main.lua]: Discussed again. Nope. It is better to have separate luacheck executable with #!/usr/bin/env tarantool shebang, which just requires luacheck.main.

File: .travis.yml [hunk 12]
```diff
@@ -16,26 +23,93 @@ matrix:
...
+        lua -e 'package.preload.lanes=error;package.path="./src/?.lua;./src/?/init.lua;"..package.path' \
+            -lluacov bin/luacheck.lua --version | grep 'Not found'
+        lua install.lua path/to/luacheck
+        mv src src2
+        path/to/luacheck/bin/luacheck spec/*.lua
+        mv src2 src
+    else
+        # Luacheck has a major.minor.patch versioning with unannotated tags.
+        # Packpack can't handle it automaticaly, because it expects
+        # major.minor annotated tags. We had to pass version by setting VERSION.
```
Comment (Totktonada) [inline .travis.yml]: packpack correctly handle major.minor.patch and likely more complex tag schemas. I think the question here is just that tags are not annotated.

Sorry for nitpicking, but I’m againts of confusing comments.

File: bin/luacheck.lua [hunk 13]
```diff
@@ -1,2 +0,0 @@
-#!/usr/bin/env lua
```
Comment (Totktonada) [inline bin/luacheck.lua]: I think we can leave it as is.

**Review by Totktonada — APPROVED** [review]
Really, last two comments from me. LGTM after fixing it (no need to re-review).

File: bin/luacheck.lua (hunk 13 above)
Comment (SudoBobo) [reply bin/luacheck.lua]: Sure, thank you for noting.

File: .travis.yml (hunk 12 above)
Comment (SudoBobo) [reply .travis.yml]: You are right. Fixed here and also in lulpeg and lrexlib.

File: rpm/luacheck.spec [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+# During package building {version} is overwritten by Packpack with
+# VERSION. It is set to major.minor.patch.number_of_commits_above_last_tag.
+# major.minor.patch tag and number of commits above are taken from the
+# github repository: https://github.com/tarantool/luacheck
+Version: 0.22.0
+Release: 1%{?dist}
+Summary: A static analyzer and a linter for Lua
+Group: Development/Tools
+License: MIT
+URL: https://github.com/mpeterv/luacheck
```
Comment (Totktonada) [inline rpm/luacheck.spec]: I think we should use our URL.

Comment (SudoBobo) [reply rpm/luacheck.spec]: Ok.

**Pull request merged**